### PR TITLE
gh-97612: Fix shell injection in get-remote-certificate.py

### PR DESCRIPTION
**Pull request by vstinner**
Fix a shell code injection vulnerability in the
get-remote-certificate.py example script. The script no longer uses a shell to run "openssl" commands. Issue reported and initial fix by Caleb Shortt.

Remove the Windows code path to send "quit" on stdin to the "openssl
s_client" command: use DEVNULL on all platforms instead.

* Issue: gh-97612

### CODE REVIEW

**Comment by vstinner** [comment]
@gpshead: Would you mind to review this fix?

The fix should also be backported to the 3.7 branch, but there is a conflict, so I will do the backport manually once this PR is merged.

cc @calebshortt

**Comment by vstinner** [comment]
> What's the point of even keeping this script? Has anyone used it in the past 12 years?

In Python 3.12, IMO it's fine to consider removing the script. But for stable Python versions (3.7-3.11), IMO we should fix the script.

I'm working on proposal to remove most Tools/scripts/ scripts, but first I have to look into these scripts to decide what to do with them, categorize them, etc. There are more than 70 scripts.

File: Tools/scripts/get-remote-certificate.py
Comment (gpshead) [inline Tools/scripts/get-remote-certificate.py:49]: on windows stdin was hooked up to a file containing "quit\n".  was that necessary?  If so, use your added subproc stdin= support from above to do that.  otherwise, get rid of the added subprob stdin= support to make this PR smaller.

Comment (vstinner) [reply Tools/scripts/get-remote-certificate.py:49]: Right, I removed the now unused stdin parameter.

This script is used to connect to a HTTPS server. Saying "quit\n" to a HTTP server makes no sense to me. Maybe on Windows, previously the command was blocked until the user wrotes something and then press ENTER? But with this change, the stdin is replaced with DEVNULL (``NUL`` device on Windows) and so it should no longer block.

First, I basically rewrote the whole script. I don't get half of the code. For example, it runs the "x509" to get the same output than the input... it doesn't make any sense to me. But I tried to keep this change as small as possible...

Then I think that we should just remove the script in the main branch :-)

My main motivation for this change is to prevent users to report the same issue later, even if the impact of the vulnerability is basically non existent since I don't imagine that anyone runs such script on a production server with inputs taken from the Internet...

**Comment by vstinner** [comment]
@gpshead: Would you mind to review the updated PR?

**Review by gpshead — APPROVED** [review]

File: Tools/scripts/get-remote-certificate.py
Comment (zware) [reply Tools/scripts/get-remote-certificate.py:49]: In that case, I'd take the current report as reason to remove the script from main, and the next report as reason to remove it from other branches.  Or just go ahead and remove it everywhere.  It's not worth fixing.

Comment (vstinner) [reply Tools/scripts/get-remote-certificate.py:49]: Since the fix is straightforward, I prefer first to fix the script in all branches.

**Comment by vstinner** [comment]
@zware: I prefer the conservative approach, first fix the vulnerability, then discuss removing outdated scripts.

File: Tools/scripts/get-remote-certificate.py
Comment (vstinner) [reply Tools/scripts/get-remote-certificate.py:49]: I created https://discuss.python.org/t/remove-outdated-tools-scripts-scripts/19571 discussion to propose removing outdated example scripts.

**Pull request merged**
